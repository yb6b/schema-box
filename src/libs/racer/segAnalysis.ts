/**
 * @file 分析分词的结果
 * @author yb6b
 */
import type { SegValue } from './segmentation'

const feel = await import('./feelData')

export interface RacerResult {
  hanzi: number // 上屏的汉字总数
  lack: number // 缺字
  lackSet: Set<string>
  commit: number // 上屏总数

  words: number // 打词的次数，rate:总上屏
  wordsChar: number // 打词的总字数，rate:总字数
  wordsDist: number[] // 不同字数的词的频数

  collision: number // 选重的次数，rate:总上屏
  collisionChar: number // 选重的总字数，rate:总字数
  collisionDist: number[] // 不同选重的词的频数

  CL: number // 上屏汉字的编码总长，rate:总上屏
  CLDist: number[] // 不同编码长度的频数

  keys: number // 总击键数
  keysDist: {
    [keyName: string]: number
  } // 不同按键的频数

  fingers: number[] // 各手指（按Fingers enum索引）的使用频率

  comb: number // 按键组合总数
  Eq: number // 总当量，注意除10
  trible: number // 三连击频数
  double: number// 二连击频数
  singleSpan: number // 小跨排频数
  multiSpan: number // 大跨排频数
  longFingerDisturb: number // 错手频数
  littleFingerDisturb: number // 小指干扰频数

  sameFingers: number // 同指组合

  LL: number// 左手到左手
  LR: number// 左手到右手
  RR: number// 右手到右手
  RL: number// 右手到左手
}

/**
 * 生成空的赛码器结果对象
 * @returns 空的赛码器结果
 */
export function createEmptyRacerResult(): RacerResult {
  return {
    hanzi: 0, // 上屏的汉字总数
    lack: 0, // 缺字
    lackSet: new Set(),
    commit: 0, // 上屏总数

    words: 0, // 打词的次数，rate:总上屏
    wordsChar: 0, // 打词的总字数，rate:总字数
    wordsDist: [], // 不同字数的词的频数

    collision: 0, // 选重的次数，rate:总上屏
    collisionChar: 0, // 选重的总字数，rate:总字数
    collisionDist: [], // 不同选重的词的频数

    CL: 0, // 上屏编码总长，rate:总上屏
    CLDist: [], // 不同编码长度的频数

    keys: 0, // 总击键数
    keysDist: {}, // 不同按键的频数

    comb: 0, // 按键组合总数
    Eq: 0, // 总当量，注意除10
    trible: 0, // 三连击频数
    double: 0, // 二连击频数
    singleSpan: 0, // 小跨排频数
    multiSpan: 0, // 大跨排频数
    longFingerDisturb: 0, // 错手
    littleFingerDisturb: 0, // 小指干扰

    sameFingers: 0, // 同指组合
    fingers: Array(9).fill(0),

    LL: 0, // 左手到左手
    LR: 0, // 左手到右手
    RR: 0, // 右手到右手
    RL: 0, // 右手到左手
  }
}

/**
 * 在数字数组中，自动递增一个元素，如果位置过大，会自动扩容
 * @param arr 要添加的数组
 * @param index 修改位置的索引号
 */
function autoIncreaseArrayElement(arr: number[], index: number) {
  if (index >= arr.length) {
    for (let i = index - arr.length; i > 0; i--)
      arr.push(0)
    arr.push(1)
  }
  else {
    arr[index]++
  }
}

/**
 * 在字-数对象中，自动递增一个元素
 * @param o 要添加的对象
 * @param k 修改位置的索引
 */
function autoIncreaseObjectElement(o: { [key: string]: number }, k: string) {
  if (k in o)
    o[k]++
  else
    o[k] = 1
}

export class SegmentAnalyser {
  private _result: RacerResult
  constructor(
    public collisionKeys = '_23456789',
    public commitLength = 4,
  ) {
    this._result = createEmptyRacerResult()
  }

  /**
   * 每次添加一个分词结果对象，分析一些数据
   * @param aSegValue 一个分词结果
   */
  public add(aSegValue: SegValue) {
    let lastCode = ''
    let lastCode2 = ''
    let thisCode = ''
    const result = this._result
    const updateCodes = () => {
      switch (thisCode.length) {
        case 0:break
        case 1:lastCode2 = lastCode
          lastCode = thisCode
          break
        default:lastCode2 = thisCode[thisCode.length - 2]
          lastCode = thisCode[thisCode.length - 1]
          break
      }
    }

    const handleEachValue = (eachValue: SegValue) => {
      if (eachValue.type === 'l') {
        result.lack++
        result.lackSet.add(eachValue.words)
      }
      else {
        // 只关于汉字
        if (eachValue.type === 'c') {
          const tempCodes = eachValue.value.code.replace('↑', '')
          const tempCollision = eachValue.value.collision
          if (tempCodes.length > this.commitLength && tempCollision === 1)
            thisCode = tempCodes
          else
            thisCode = tempCodes + this.collisionKeys[tempCollision - 1]

          const wordsLength = [...eachValue.words].length
          result.hanzi += wordsLength
          result.commit++

          if (wordsLength > 1) {
            result.words++
            result.wordsChar += wordsLength
          }
          autoIncreaseArrayElement(result.wordsDist, wordsLength - 1)

          result.CL += thisCode.length
          autoIncreaseArrayElement(result.CLDist, thisCode.length - 1)

          const tmpCollision = eachValue.value.collision
          if (tmpCollision > 1) {
            result.collision++
            result.collisionChar += wordsLength
          }
          autoIncreaseArrayElement(result.collisionDist, tmpCollision - 1)
        }
        else {
          thisCode = eachValue.code.replace('↑', '')
        }
        // 只要有编码就行
        result.keys += thisCode.length
        // 每个按键
        for (let j = 0; j < thisCode.length; j++) {
          const tmpCode = thisCode[j]
          autoIncreaseObjectElement(result.keysDist, tmpCode)
          result.fingers[feel.keyToFinger[tmpCode]]++
        }
      }
    }

    const handleCombsCode = () => {
      const expandCode = lastCode + thisCode
      for (let i = 1; i < expandCode.length; i++) {
        result.comb++
        const comb = expandCode[i - 1] + expandCode[i]
        const feelMagic = feel.keyPairData[comb]
        result.Eq += feel.getEquivalent(feelMagic)
        if (feel.isDoubleHit(feelMagic))
          result.double++
        if (feel.isSingleSpan(feelMagic))
          result.singleSpan++
        if (feel.isMultiSpan(feelMagic))
          result.multiSpan++
        if (feel.isLongFinersDisturb(feelMagic))
          result.longFingerDisturb++
        if (feel.isLittleFinersDisturb(feelMagic))
          result.littleFingerDisturb++

        const firstFinger = feel.keyToFinger[expandCode[i - 1]]
        const secondFinger = feel.keyToFinger[expandCode[i]]
        if (firstFinger === secondFinger)
          result.sameFingers++

        const isFirstHandLeft = feel.isLeftHand(firstFinger)
        const isSecondHandLeft = feel.isLeftHand(secondFinger)
        if (isFirstHandLeft) {
          if (isSecondHandLeft)
            result.LL++
          else
            result.LR++
        }
        else {
          if (isSecondHandLeft)
            result.RL++
          else
            result.RR++
        }
      }

      const expandCode2 = lastCode2 + expandCode
      for (let i = 2; i < expandCode2.length; i++) {
        if (expandCode2[i - 2] === expandCode2[i - 1] && expandCode2[i - 1] === expandCode2[i])
          result.trible++
      }
    }

    handleEachValue(aSegValue)
    handleCombsCode()
    updateCodes()
  }

  /**
   * 所有的分词添加完成后，调用之，setter 函数会修正数据。
   */
  get result() {
    this._result.Eq /= 10
    return this._result
  }
}
