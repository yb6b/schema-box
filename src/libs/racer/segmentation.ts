/**
 * @file 按照字典树，拆分文章成一条条词语
 * @author yb6b
 */

import { PUNCTUATIONS } from '../schema/keys'
import type { TrieTreeValue, UniqueTrieTree } from './uniqueTrieTree'

interface SegValueCn {
  type: 'c'
  index: number
  words: string
  value: TrieTreeValue
}

interface SegValuePunc {
  type: 'p'
  index: number
  words: string
  code: string
}

interface SegValueEn {
  type: 'e'
  index: number
  words: string
  code: string
}

interface SegValueUni {
  type: 'u'
  index: number
  words: string
  code: string
}

interface SegValueLack {
  type: 'l'
  index: number
  words: string
}

export type SegValue = SegValueCn | SegValueEn | SegValueLack | SegValuePunc | SegValueUni

/**
 * 用贪婪算法拆分文章。优先用最长的词匹配。
 * @param src 要分词的文章，请确保文章不是空字符串
 * @param tree 方案的字典树
 */
export function segmentGreedily(src: string, tree: UniqueTrieTree): SegValue[] {
  const maxIndex = src.length - 1
  const result: SegValue[] = []
  let beginIndex = 0
  let endIndex = 0
  let bestWords: string
  let bestValue: TrieTreeValue

  const maybeSignalButZh = () => {
    const getCode = (dict: any) => {
      const ei = endIndex
      const bi = beginIndex
      const d = dict
      const tmpResult: string[] = new Array(ei - bi)
      for (let i = bi; i < ei; i++)
        tmpResult[i - bi] = d[src[i]]
      return tmpResult.join('')
    }
    const twoLetterPunctuation = (punc: string) => {
      const puncCode = PUNCTUATIONS.cn[punc]
      if (src[beginIndex + 1] === punc) {
        result.push({ type: 'p', words: punc + punc, index: beginIndex, code: puncCode + puncCode })
        beginIndex += 2
      }
      else {
        result.push({ type: 'p', words: punc, index: beginIndex, code: puncCode })
        beginIndex += 1
      }
    }
    const firstWord = src[beginIndex]
    endIndex = beginIndex
    // 是中文标点？
    if (firstWord in PUNCTUATIONS.cn) {
      switch (firstWord) {
        case '…': twoLetterPunctuation('…')
          break
        case '—': twoLetterPunctuation('—')
          break
        default:result.push({ type: 'p', words: firstWord, index: beginIndex, code: PUNCTUATIONS.cn[firstWord] })
          beginIndex++
          break
      }
    }
    // 中英通用的标点？
    else if (firstWord in PUNCTUATIONS.uni) {
      while (src[endIndex] in PUNCTUATIONS.uni && endIndex <= maxIndex) endIndex++
      result.push({ type: 'u', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.uni) })
      beginIndex = endIndex
    }
    // 英文符号？
    else if (firstWord in PUNCTUATIONS.en) {
      while (src[endIndex] in PUNCTUATIONS.en && endIndex <= maxIndex) endIndex++
      result.push({ type: 'e', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.en) })
      beginIndex = endIndex
    }
    // 只能是缺字了
    else {
      result.push({ type: 'l', words: firstWord, index: beginIndex })
      beginIndex += firstWord.length
    }
  }
  // 如果第一轮就挂了，需要best*数据回退
  const scanTree = () => {
    // 先处理第一次运行
    while (true) {
      // 到文末
      if (maxIndex < endIndex) {
        result.push({ type: 'c', index: beginIndex, words: bestWords, value: bestValue })
        beginIndex += bestWords.length
        return
      }
      const tmpWords = src.slice(beginIndex, endIndex)
      const tmpValue = tree.data.get(tmpWords)
      if (tmpValue) {
        // 到最长
        if (tmpValue.prefixeeLen === tmpWords.length) {
          result.push({ type: 'c', index: beginIndex, words: tmpWords, value: tmpValue })
          beginIndex = endIndex
          return
        }
        // 否则仍需要迭代，用几个 best* 暂存结果
        bestWords = tmpWords
        bestValue = tmpValue
        endIndex = beginIndex + tmpValue.prefixeeLen
      }
      // 找不到了
      else {
        result.push({ type: 'c', index: beginIndex, words: bestWords, value: bestValue })
        beginIndex += bestWords.length
        return
      }
    }
  }

  // 每次循环生成添加一个 Seg
  do {
    const firstWord = String.fromCodePoint(src.codePointAt(beginIndex) as number)
    const firstWordValue = tree.data.get(firstWord)
    if (firstWordValue) {
      // 数据库里只收录了该单字，不可能是空穴
      if (firstWord.length === firstWordValue.prefixeeLen) {
        result.push({ type: 'c', index: beginIndex, words: firstWord, value: firstWordValue })
        beginIndex += firstWord.length
        continue
      }

      bestValue = firstWordValue
      bestWords = firstWord
      endIndex = beginIndex + firstWordValue.prefixeeLen

      // 第一个字不是空穴
      if (firstWordValue.code) {
        scanTree()
      }
      // 第一个字是空穴
      else {
        const secondWord = src.slice(beginIndex, endIndex)
        const secondWordValue = tree.data.get(secondWord)
        // 能匹配更长的，必然不是空首字
        if (secondWordValue) {
          // 数据库里只收录了词，不可能是空穴
          if (secondWord.length === secondWordValue.prefixeeLen) {
            result.push({ type: 'c', index: beginIndex, words: secondWord, value: secondWordValue })
            beginIndex += secondWord.length
            continue
          }
          bestWords = secondWord
          bestValue = secondWordValue
          endIndex = beginIndex + secondWordValue.prefixeeLen
          scanTree()
        }
        // 首字是空的，有可能是标点等
        else {
          maybeSignalButZh()
        }
      }
    }
    // 未收录或空穴的，是缺字或别的情况
    else {
      maybeSignalButZh()
    }
  } while (beginIndex <= maxIndex)

  return result
}

export function* generateSegmentGreedily(src: string, tree: UniqueTrieTree) {
  const maxIndex = src.length - 1
  let beginIndex = 0
  let endIndex = 0
  let bestWords: string
  let bestValue: TrieTreeValue

  // 每次循环生成添加一个 Seg
  do {
    const firstWord = String.fromCodePoint(src.codePointAt(beginIndex) as number)
    const firstWordValue = tree.data.get(firstWord)
    if (firstWordValue) {
      // 数据库里只收录了该单字，不可能是空穴
      if (firstWord.length === firstWordValue.prefixeeLen) {
        yield { type: 'c', index: beginIndex, words: firstWord, value: firstWordValue } as SegValueCn
        beginIndex += firstWord.length
        continue
      }

      bestValue = firstWordValue
      bestWords = firstWord
      endIndex = beginIndex + firstWordValue.prefixeeLen

      // 第一个字不是空穴
      if (firstWordValue.code) {
        yield scanTree()
      }
      // 第一个字是空穴
      else {
        const secondWord = src.slice(beginIndex, endIndex)
        const secondWordValue = tree.data.get(secondWord)
        // 能匹配更长的，必然不是空首字
        if (secondWordValue) {
          // 数据库里只收录了词，不可能是空穴
          if (secondWord.length === secondWordValue.prefixeeLen) {
            yield { type: 'c', index: beginIndex, words: secondWord, value: secondWordValue } as SegValueCn
            beginIndex += secondWord.length
            continue
          }
          bestWords = secondWord
          bestValue = secondWordValue
          endIndex = beginIndex + secondWordValue.prefixeeLen
          yield scanTree()
        }
        // 首字是空的，有可能是标点等
        else {
          yield maybeSignalButZh()
        }
      }
    }
    // 未收录或空穴的，是缺字或别的情况
    else {
      yield maybeSignalButZh()
    }
  } while (beginIndex <= maxIndex)

  function maybeSignalButZh(): SegValue {
    const getCode = (dict: any) => {
      const ei = endIndex
      const bi = beginIndex
      const d = dict
      const tmpResult: string[] = new Array(ei - bi)
      for (let i = bi; i < ei; i++)
        tmpResult[i - bi] = d[src[i]]
      return tmpResult.join('')
    }
    const twoLetterPunctuation = (punc: string): SegValuePunc => {
      const puncCode = PUNCTUATIONS.cn[punc]
      if (src[beginIndex + 1] === punc) {
        const tmp: SegValuePunc = { type: 'p', words: punc + punc, index: beginIndex, code: puncCode + puncCode }
        beginIndex += 2
        return tmp
      }
      else {
        const tmp: SegValuePunc = { type: 'p', words: punc, index: beginIndex, code: puncCode }
        beginIndex += 1
        return tmp
      }
    }
    let tmp: SegValue
    const firstWord = src[beginIndex]
    endIndex = beginIndex
    // 是中文标点？
    if (firstWord in PUNCTUATIONS.cn) {
      switch (firstWord) {
        case '…': return twoLetterPunctuation('…')
        case '—': return twoLetterPunctuation('—')
        default: tmp = { type: 'p', words: firstWord, index: beginIndex, code: PUNCTUATIONS.cn[firstWord] }
          beginIndex++
      }
    }
    // 中英通用的标点？
    else if (firstWord in PUNCTUATIONS.uni) {
      while (src[endIndex] in PUNCTUATIONS.uni && endIndex <= maxIndex) endIndex++

      tmp = { type: 'u', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.uni) }
      beginIndex = endIndex
    }
    // 英文符号？
    else if (firstWord in PUNCTUATIONS.en) {
      while (src[endIndex] in PUNCTUATIONS.en && endIndex <= maxIndex) endIndex++
      tmp = { type: 'e', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.en) }
      beginIndex = endIndex
    }
    // 只能是缺字了
    else {
      tmp = { type: 'l', words: firstWord, index: beginIndex }
      beginIndex += firstWord.length
    }
    return tmp
  }

  // 如果第一轮就挂了，需要best*数据回退
  function scanTree() {
    // 先处理第一次运行
    while (true) {
      let tmp: SegValue
      // 到文末
      if (maxIndex < endIndex) {
        tmp = { type: 'c', index: beginIndex, words: bestWords, value: bestValue }
        beginIndex += bestWords.length
        return tmp
      }
      const tmpWords = src.slice(beginIndex, endIndex)
      const tmpValue = tree.data.get(tmpWords)
      if (tmpValue) {
        // 到最长
        if (tmpValue.prefixeeLen === tmpWords.length) {
          tmp = { type: 'c', index: beginIndex, words: tmpWords, value: tmpValue }
          beginIndex = endIndex
          return tmp
        }
        // 否则仍需要迭代，用几个 best* 暂存结果
        bestWords = tmpWords
        bestValue = tmpValue
        endIndex = beginIndex + tmpValue.prefixeeLen
      }
      // 找不到了
      else {
        tmp = { type: 'c', index: beginIndex, words: bestWords, value: bestValue }
        beginIndex += bestWords.length
        return tmp
      }
    }
  }
}
