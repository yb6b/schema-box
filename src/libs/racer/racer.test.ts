import { describe, expect, it } from 'vitest'
import { CollisionCalculator, UniqueTrieTree, getJisuCollision } from './uniqueTrieTree'
import { generateSegmentGreedily, segmentGreedily } from './segmentation'
import { SegmentAnalyser } from './segAnalysis'

export function addTestTree(i: string, tree: UniqueTrieTree) {
  i = i.trim()
  let lineno = 0
  for (const a of i.split('\n')) {
    lineno++
    const [words, codes] = a.trim().split(' ')
    tree.add(words, { code: codes, line: lineno, collision: 1, prefixeeLen: 0 })
  }
}

describe('通常情况下', () => {
  it('短码优先加词', () => {
    const u = new UniqueTrieTree()
    addTestTree(`
      好 abb
      好 aa`, u)
    expect(u.get('好')?.code).eq('aa')
  })

  it('行数最短优先加词', () => {
    const u = new UniqueTrieTree(true)
    addTestTree(`
      好 abb
      好 aa`, u)
    expect(u.get('好')?.code).eq('abb')
  })

  it('有空穴', () => {
    const u = new UniqueTrieTree(true)
    addTestTree(`
      海 a
      海上生明月 aaaa`, u)
    expect(u.get('海上')).eq(undefined)
    expect(u.data.get('海上')).toEqual({ code: '', collision: 0, line: 0, prefixeeLen: 5 })

    expect(u.data.get('海')).toEqual({ code: 'a', collision: 1, line: 1, prefixeeLen: 5 })
    u.delete('海')
    expect(u.data.get('海')).toEqual({ code: '', collision: 1, line: 1, prefixeeLen: 5 })

    u.delete('海上生明月')
    expect(u.data.get('海')).toMatchInlineSnapshot('undefined')
  })
})

describe('计算重码', () => {
  it('正常加词', () => {
    const c = new CollisionCalculator()
    expect(c.add('a')).toBe(1)
    expect(c.add('a')).toBe(2)
    c.delete('a')
    expect(c.get('a')).toBe(1)
    c.delete('a')
    expect(c.get('a')).toBe(0)
    c.delete('a')
    expect(c.get('a')).toBe(0)
  })
})

describe('极速赛码表的重码', () => {
  it('正常', () => {
    expect(getJisuCollision('').collision).toBe(1)
    expect(getJisuCollision('234').collision).toBe(234)
    expect(getJisuCollision('aaa234').collision).toBe(234)
    expect(getJisuCollision('aaa2_').collision).toBe(1)
  })
})

describe('贪婪分词', () => {
  const t = new UniqueTrieTree()
  addTestTree(`
好 a
好的 b
好好睡觉 C
睡觉 d
xo e`, t)
  it('只有汉字', () => {
    expect(segmentGreedily('好好的睡觉', t)).toMatchSnapshot()
  })
  it('含有英文', () => {
    expect(segmentGreedily('好xo觉bbbb', t)).toMatchSnapshot()
  })
  it('含有空格', () => {
    expect(segmentGreedily('好  好  ', t)).toMatchSnapshot()
  })
  it('含有标点', () => {
    expect(segmentGreedily('好的！', t)).toMatchSnapshot()
  })
  it('含有大写字母', () => {
    expect(segmentGreedily('好好睡觉！', t)).toMatchSnapshot()
  })
})

describe('贪婪分词生成器', () => {
  const t = new UniqueTrieTree()
  addTestTree(`
好 a
好的 b
好好睡觉 c
睡觉 d
xo e`, t)
  it('只有汉字', () => {
    expect([...generateSegmentGreedily('好好的睡觉', t)]).toMatchSnapshot()
  })
  it('含有英文', () => {
    expect([...generateSegmentGreedily('好xo觉bbbb', t)]).toMatchSnapshot()
  })
  it('含有空格', () => {
    expect([...generateSegmentGreedily('好  好  ', t)]).toMatchSnapshot()
  })
  it('含有标点', () => {
    expect([...generateSegmentGreedily('好的！', t)]).toMatchSnapshot()
  })
})

describe('分析', () => {
  const t = new UniqueTrieTree()
  addTestTree(`
好 a
好的 b
好好睡觉 c
睡觉 d
xo e`, t)
  const seg = [...generateSegmentGreedily('好觉b bbb睡觉24,', t)]
  it('常规', () => {
    expect(seg).toMatchSnapshot()
    const m = new SegmentAnalyser()

    seg.forEach((e) => {
      m.add(e)
    })
    expect(m.result).toMatchSnapshot()
  })
})
