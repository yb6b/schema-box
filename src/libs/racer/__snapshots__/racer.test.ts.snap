// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`分析 > 常规 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 1,
    "type": "l",
    "words": "觉",
  },
  {
    "code": "b",
    "index": 2,
    "type": "e",
    "words": "b",
  },
  {
    "code": "_",
    "index": 3,
    "type": "u",
    "words": " ",
  },
  {
    "code": "bbb",
    "index": 4,
    "type": "e",
    "words": "bbb",
  },
  {
    "index": 7,
    "type": "c",
    "value": {
      "code": "d",
      "collision": 1,
      "line": 4,
      "prefixeeLen": 2,
    },
    "words": "睡觉",
  },
  {
    "code": "24",
    "index": 9,
    "type": "u",
    "words": "24",
  },
  {
    "code": ",",
    "index": 11,
    "type": "e",
    "words": ",",
  },
]
`;

exports[`分析 > 常规 2`] = `
{
  "CL": 4,
  "CLDist": [
    0,
    2,
  ],
  "Eq": 6.8,
  "LL": 3,
  "LR": 2,
  "RL": 0,
  "RR": 0,
  "collision": 0,
  "collisionChar": 0,
  "collisionDist": [
    2,
  ],
  "comb": 5,
  "commit": 2,
  "double": 2,
  "fingers": [
    5,
    1,
    1,
    1,
    0,
    1,
    0,
    0,
    3,
  ],
  "hanzi": 3,
  "keys": 12,
  "keysDist": {
    ",": 1,
    "2": 1,
    "4": 1,
    "_": 3,
    "a": 1,
    "b": 4,
    "d": 1,
  },
  "lack": 1,
  "lackSet": Set {
    "觉",
  },
  "littleFingerDisturb": 0,
  "longFingerDisturb": 0,
  "multiSpan": 0,
  "sameFingers": 2,
  "singleSpan": 0,
  "trible": 1,
  "words": 1,
  "wordsChar": 2,
  "wordsDist": [
    1,
    1,
  ],
}
`;

exports[`贪婪分词 > 只有汉字 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 1,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 2,
    "type": "l",
    "words": "的",
  },
  {
    "index": 3,
    "type": "c",
    "value": {
      "code": "d",
      "collision": 1,
      "line": 4,
      "prefixeeLen": 2,
    },
    "words": "睡觉",
  },
]
`;

exports[`贪婪分词 > 含有大写字母 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "C",
      "collision": 1,
      "line": 3,
      "prefixeeLen": 4,
    },
    "words": "好好睡觉",
  },
  {
    "code": "↑1",
    "index": 4,
    "type": "p",
    "words": "！",
  },
]
`;

exports[`贪婪分词 > 含有标点 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 1,
    "type": "l",
    "words": "的",
  },
  {
    "code": "↑1",
    "index": 2,
    "type": "p",
    "words": "！",
  },
]
`;

exports[`贪婪分词 > 含有空格 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "code": "__",
    "index": 1,
    "type": "u",
    "words": "  ",
  },
  {
    "index": 3,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "code": "__",
    "index": 4,
    "type": "u",
    "words": "  ",
  },
]
`;

exports[`贪婪分词 > 含有英文 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 1,
    "type": "c",
    "value": {
      "code": "e",
      "collision": 1,
      "line": 5,
      "prefixeeLen": 2,
    },
    "words": "xo",
  },
  {
    "index": 3,
    "type": "l",
    "words": "觉",
  },
  {
    "code": "bbbb",
    "index": 4,
    "type": "e",
    "words": "bbbb",
  },
]
`;

exports[`贪婪分词生成器 > 只有汉字 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 1,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 2,
    "type": "l",
    "words": "的",
  },
  {
    "index": 3,
    "type": "c",
    "value": {
      "code": "d",
      "collision": 1,
      "line": 4,
      "prefixeeLen": 2,
    },
    "words": "睡觉",
  },
]
`;

exports[`贪婪分词生成器 > 含有标点 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 1,
    "type": "l",
    "words": "的",
  },
  {
    "code": "↑1",
    "index": 2,
    "type": "p",
    "words": "！",
  },
]
`;

exports[`贪婪分词生成器 > 含有空格 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "code": "__",
    "index": 1,
    "type": "u",
    "words": "  ",
  },
  {
    "index": 3,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "code": "__",
    "index": 4,
    "type": "u",
    "words": "  ",
  },
]
`;

exports[`贪婪分词生成器 > 含有英文 1`] = `
[
  {
    "index": 0,
    "type": "c",
    "value": {
      "code": "a",
      "collision": 1,
      "line": 1,
      "prefixeeLen": 4,
    },
    "words": "好",
  },
  {
    "index": 1,
    "type": "c",
    "value": {
      "code": "e",
      "collision": 1,
      "line": 5,
      "prefixeeLen": 2,
    },
    "words": "xo",
  },
  {
    "index": 3,
    "type": "l",
    "words": "觉",
  },
  {
    "code": "bbbb",
    "index": 4,
    "type": "e",
    "words": "bbbb",
  },
]
`;
